src/click/core.py

```diff
@@ -2528,17 +2534,17 @@ def handle_parse_result(
             # We skip adding the value if it was previously set by another parameter
             # targeting the same variable name. This prevents parameters competing for
             # the same name to override each other.
-            and self.name not in ctx.params
+            and (self.name not in ctx.params or ctx.params[self.name] is UNSET)
         ):
             # Click is logically enforcing that the name is None if the parameter is
             # not to be exposed. We still assert it here to please the type checker.
             assert self.name is not None, (
                 f"{self!r} parameter's name should not be None when exposing value."
             )
-            # Normalize UNSET values to None, as we're about to pass them to the
-            # command function and move them to the pure-Python realm of user-written
-            # code.
-            ctx.params[self.name] = value if value is not UNSET else None
+            # Do not normalize UNSET values to None at this point
+            # It needs to be done, but doing so too early impacts multiple parameters
+            # targeting the same variable name
+            ctx.params[self.name] = value
```

Comment:
Re-checking my changes, I'm not sure if I should keep this comment. Very open to a nudge about rewriting or even just removing it.